packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,174 @@
+
+const useUtilityClasses = (styleProps) => {
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+  };
+
+  return composeClasses(slots, getMasonryItemUtilityClass, classes);
+};
+
+export const style = ({ styleProps, theme }) => {
+  let styles = {
+    width: '100%',
+    '& > *': {
+      // all contents should have a width of 100%
+      width: '100%',
+      boxSizing: 'inherit',
+    },
+    visibility: styleProps.height ? 'visible' : 'hidden',
+    gridColumnEnd: `span ${styleProps.columnSpan}`,
+    boxSizing: 'inherit',
+  };
+
+  const base = {};
+  Object.keys(theme.breakpoints.values).forEach((breakpoint) => {
+    if (styleProps.spacing[breakpoint] != null) {
+      base[breakpoint] = true;
+    }
+  });
+  const spacingValues = resolveBreakpointValues({ values: styleProps.spacing, base });
+  const transformer = createUnarySpacing(theme);
+  const styleFromPropValue = (propValue) => {
+    const gap = styleProps.height ? Number(getValue(transformer, propValue).replace('px', '')) : 0;
+    // for lazy-loaded images to load properly, they should take up space greater than 1px
+    // taking into account a row gap of 1px, rowSpan should at least be 3
+    const rowSpan = styleProps.height ? Math.ceil(styleProps.height + gap) : 3;
+    return {
+      gridRowEnd: `span ${rowSpan}`,
+      paddingBottom: gap - 1,
+    };
+  };
+  styles = { ...styles, ...handleBreakpoints({ theme }, spacingValues, styleFromPropValue) };
+
+  return styles;
+};
+
+const MasonryItemRoot = styled('div', {
+  name: 'MuiMasonryItem',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(style);
+
+const MasonryItem = React.forwardRef(function MasonryItem(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const masonryItemRef = React.useRef(null);
+
+  const { spacing = 1 } = React.useContext(MasonryContext);
+  const { children, className, component = 'div', columnSpan = 1, height, ...other } = props;
+  const [styleProps, setStyleProps] = React.useState({
+    ...props,
+    spacing,
+    columnSpan,
+    height: height < 0 ? 0 : height, // MasonryItems to which negative or zero height is passed will be hidden
+  });
+
+  const classes = useUtilityClasses(styleProps);
+
+  React.useEffect(() => {
+    // If height is passed by user, ResizeObserver is not used.
+    if (height !== undefined) {
+      setStyleProps({
+        ...styleProps,
+        spacing,
+        columnSpan,
+        height: height < 0 ? 0 : height,
+      });
+
+      return true;
+    }
+    const resizeObserver = new ResizeObserver(([item]) => {
+      setStyleProps({
+        ...styleProps,
+        spacing,
+        columnSpan,
+        height: item.contentRect.height,
+      });
+    });
+    const handleError = () => {
+      const src = masonryItemRef.current.firstChild.src;
+      masonryItemRef.current.firstChild.src = src;
```

Comment:
What are these lines doing?

Reply:
In case of images, if they fail loading, this will try to load them again. I added a timestamp to prevent this being ignored from cache: https://github.com/mui-org/material-ui/pull/27439/commits/4913ad805b559c00d19f792df3365718fa61fec7